group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -3500,6 +3503,59 @@ public void maybeDeleteGroup(String groupId, List<Record> records) {
         }
     }
 
+    /**
+     * A group can be upgraded offline if it's a classic group and empty.
+     *
+     * @param groupId The group to be validated.
+     * @return true if the offline upgrade is valid.
+     */
+    private boolean validateOfflineUpgrade(String groupId) {
+        Group group = groups.get(groupId);
+
+        if (group == null || group.type() == CONSUMER) {
+            return false;
+        }
+
+        ClassicGroup classicGroup = (ClassicGroup) group;
+        if (!classicGroup.isEmpty()) {
+            return false;
+        } else {
+            return true;
+        }
+    }
+
+    /**
+     * Upgrade the empty classic group to a consumer group if it's valid.
+     *
+     * @param groupId       The group id to be updated.
+     * @param records       The list of records to delete the classic group and create the consumer group.
+     * @param isSimpleGroup The boolean indicating whether the group to be updated is a simple group.
+     */
+    public void maybeUpgradeEmptyGroup(String groupId, List<Record> records, boolean isSimpleGroup) {
```

Comment:
In this particular case where the group is empty and a new member joins, my understanding is that we only need to delete the previous group. The new group will be automatically created with the new member. Am I missing something?

Reply:
I agree with it in the case of upgrading if a new member joining the empty group, but for the simple group we still need to create the new group. Is that correct?